### PR TITLE
Improved documentation of GBPTree

### PR DESCRIPTION
**Pull request by burqen**

### CODE REVIEW

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java [hunk 1]
```diff
@@ -18,13 +18,135 @@
...
  * <p>
- * To create an index with a custom layout (type of key/value), implement a custom
- * {@link org.neo4j.index.internal.gbptree.Layout}.
+ * To create an index with a custom layout (type of key:value), implement a custom {@link org.neo4j.index.internal.gbptree.Layout}.
  * <p>
  * See https://en.wikipedia.org/wiki/B%2B_tree
+ * <h1>Entries move from left to right</h1>
+ * When we make structural changes to the tree we are only allowed to move entries to the right, never to the left.
+ * {@link org.neo4j.index.internal.gbptree.SeekCursor Readers} rely on this assumption being true to read correct
+ * entries without missing anything. This is also way backwards seek needs to be done with extra care. Moving entries
```
Comment (tinwelint) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java]: Typo: `...This is also way backwards seek needs...` --> `...This is also why backwards seek needs...`

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java [hunk 2]
```diff
@@ -18,13 +18,135 @@
...
+ * <p>
+ * BUBBLE KEY and BUBBLE SOURCE - Range of leaf A has been removed from range of subtree A and added to range of
+ * subtree B. The key that divided range of leaf A from the rest of the range of subtree A sits to the very right
+ * in leaf A's subtree. This means to the very right of leaf A's parent, or parent's parent if parent has no keys,
+ * etc. There is no internal key that sits further "to the right" than this one in leaf A's subtree. From the
+ * common ancestor's perspective this is "the rightmost internal key in the subtree to the left of old splitter key".
+ * Call this key the "bubble key" and the internal node where we found it the "bubble source". The bubble key is the
+ * new splitter key that divide subtree A from subtree B and we need to replace old splitter key with this.
+ * <p>
+ * Replacing the splitter key is not enough however. We also need to remove the remove bubble key from the bubble
```
Comment (tinwelint) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java]: One `remove` too many `We also need to remove the remove bubble key from`

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java [hunk 3]
```diff
@@ -18,13 +18,135 @@
...
+ * to leaf B, never the other way around. This means we move keys from range of leaf A to range of leaf B. The process
+ * of propagating this structure change is very similar to what we do on merge. We need to replace the splitter key
+ * with the new leftmost key in leaf B. We do this with a simple replace instead of using the bubble strategy.
+ * <p>
+ * SPLIT
+ * <p>
+ * We split a node A into A' and B'. To reflect this in parent we need to insert a new child pointer (to B') and a
+ * splitter key that divide range A' from range B'. This key is the leftmost in B'. The splitter key is passed up to
+ * parent together with new child pointer to be inserted. Note that even though we pass the splitter key up, we also
+ * keep it i B'.
```
Comment (tinwelint) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java]: `keep it i B'` --> `keep it in B'`

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java [hunk 4]
```diff
@@ -18,13 +18,135 @@
...
+ * <p>
+ * When inserting new splitter key and child pointer in parent node it can cause an overflow such that we need to
+ * split parent as well. Splitting an internal node is done in a similar fashion to splitting a leaf node with the
+ * slight difference that we do not keep the new splitter key among the two internal nodes but instead just pass i
```
Comment (tinwelint) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/package-info.java]: `...but instead just pass i` --> `...but instead just pass it`

**Review by tinwelint — COMMENTED** [review]
Great commentary 👏 , makes these internals much more understandable.

Just saw some minor types, which I highlighted in comments.

**Comment by burqen** [comment]
@tinwelint Good catches. Updated 👍

**Review by tinwelint — APPROVED** [review]

**Pull request merged**